Source/Fuse.Reactive.Expressions/LookUp.uno

```diff
@@ -0,0 +1,213 @@
+				_lu = lu;
+				_colSub = _lu.Index.Subscribe(context, this);
+				_indexSub = _lu.Collection.Subscribe(context, this);
+			}
+
+			bool _hasCollection;
+			object _collection;
+			bool _hasIndex;
+			object _index;
+
+			public void OnNewData(IExpression source, object value)
+			{
+				if (_lu == null) return;
+				if (source == _lu.Index) NewIndex(value);
+				if (source == _lu.Collection) NewCollection(value);
+			}
+
+			IDisposable _indexForwarder;
+			void NewIndex(object ind)
+			{
+				DisposeIndexSub();
+
+				var obs = ind as IObservable;
+				if (obs != null)
+				{
+					_indexForwarder = new ValueForwarder(obs, this);
+				}
+				else
+				{
+					_index = ind;
+					_hasIndex = true;
+					ResultChanged();
+				}
+			}
+
+			IDisposable _diag;
+
+			public void SetDiagnostic(string message, IExpression source)
+			{
+				ClearDiagnostic();
+				_diag = Diagnostics.ReportTemporalUserWarning(message, source);
+			}
+
+			public void ClearDiagnostic()
+			{
+				if (_diag != null)
+				{
+					_diag.Dispose();
+					_diag = null;
+				}
+			}
+
+			void ValueForwarder.IValueListener.NewValue(object value)
+			{
+				_index = value;
+				_hasIndex = true;
+				ResultChanged();
+			}
+
+			void DisposeIndexSub()
+			{
+				if (_indexForwarder != null)
+				{
+					_indexForwarder.Dispose();
+					_indexForwarder = null;
+				}
+			}
+
+			IDisposable _colObservableSub;
+			void NewCollection(object col)
+			{
+				_collection = col;
+				_hasCollection = true;
+
+				DisposeCollectionObservableSub();
+
+				var obs = col as IObservable;
+				if (obs != null) 
+					_colObservableSub = obs.Subscribe(this);
+
+				ResultChanged();
+			}
+
+			void DisposeCollectionObservableSub()
+			{
+				if (_colObservableSub != null)
+				{
+					_colObservableSub.Dispose();
+					_colObservableSub = null;
+				}
+			}
+
+			void ResultChanged()
+			{
+				if (_listener == null) return;
+				ClearDiagnostic();
+
+				if (!_hasIndex) return;
+				if (!_hasCollection) return;
+				if (_index == null || _collection == null) PushNewData(null);
+
+				var arr = _collection as IArray;
+				if (arr != null)
+				{
+					int index = 0;
+					try
+					{
+						index = Marshal.ToInt(_index);
+					}
+					catch (MarshalException me)
+					{
+						SetDiagnostic("Index must be a number: " + me.Message, _lu.Index);
+						return;
+					}
+
+					if (index >= 0 && index < arr.Length)
+					{
+						PushNewData(arr[index]);
+					}
+					else
+					{
+						SetDiagnostic("Index was outside the bounds of the array", _lu.Index);
+					}
+					return;
+				}
+
+				var obj = _collection as IObject;
+				if (obj != null)
+				{
+					var key = _index.ToString();
+					if (obj.ContainsKey(key))
+					{
+						PushNewData(obj[key]);
+					}
+					else
+					{
+						SetDiagnostic("Object does not contain the given key '" + key + "'", _lu.Index);
+					}
+					return;
+				}
+
+				SetDiagnostic("Unsupported collection type: " + _collection, _lu.Collection);
```

Comment:
Perhaps this should include some more information... How about "Trying to index an unsupported collection type: "... instead? Just saying "unsupported" is a bit hard to know exactly what went wrong...